### ISSUE TITLE
Connection issues with NetKVM + Windows Server 2022 combination

### ISSUE THREAD

**Issue by olljanat** [virtio-win/kvm-guest-drivers-windows#583]
Looks that something have changed on Windows Server 2022 (preview build 20344) comparing to Windows Server 2019 (version 1809) because NetKVM driver does not works correctly on it.

Issues which I have seen are that:
- RDP connection cannot used as it gets stuck after a while (TightVNC works)
- Some DNS requests fails and it looks to be happening especially with short names so I guess that problem is at least with small UDP packages:
![image](https://user-images.githubusercontent.com/6213926/118983277-edb7cf00-b984-11eb-88e4-77043ba28b64.png)

If I switch to Intel e1000 emulated NIC then those issues disappear:
![image](https://user-images.githubusercontent.com/6213926/118983598-45eed100-b985-11eb-8145-cbbf6e6bdd0a.png)

It is also worth to mention that I run this test VM on Nutanix AHV platform so if others are not able to reproduce issue then it might be platform specific too (even when it looks like driver + Windows compatibility issue for me). I tested with both latest version of Nutanix packaged driver and 0.1.190 from here and both have same issue.

PS. I reported this one also to Windows Server Insiders forum https://techcommunity.microsoft.com/t5/windows-server-insiders/connection-issues-with-netkvm-windows-server-2022-combination/m-p/2371881

**Comment by sb-ntnx** [comment]
Hello Olli, 

Thank you for reporting this issue. I have had a quick look and indeed can reproduce the same behavior with Windows Server 2022 (preview build 20344), even though the previous preview build does not have this issue. 

I found that disabling Offload.Tx.Checksum mitigates the issue. Therefore, may I ask you to verify the same? You can disable Tx Checksum offloading via Device Manager -> Network Adapters -> Open properties of "Nutanix VirtIO Ethernet Adapter" -> Find "Offload.Tx.Checksum" parameter in the list on 'Advanced' tab and then change "Value" to "Disabled" or "TCP(v4)". 

BR,
Sergey

**Comment by olljanat** [comment]
@sb-ntnx Yes, that looks to be valid workaround. Thanks, I will use that then on my lab.

**Comment by leidwang** [comment]
Try to reproduce it on my host, but failed.

qemu version:qemu-kvm-6.0.0-16.module+el8.5.0+10848+2dccc46d.x86_64

The qemu command line:
-device virtio-net-pci,mac=9a:30:66:51:5b:46,id=idCkSbbZ,mq=on,netdev=idoK0011,bus=pcie-root-port-3,addr=0x0  \
-netdev tap,id=idoK0011,vhost=on,queues=6 \

Related files:
Windows_InsiderPreview_Server_vNext_en-us_20344.iso
virtio-win-prewhql-0.1-199.iso

![win2022](https://user-images.githubusercontent.com/63003119/119456980-c50d4c00-bd6d-11eb-8712-a0b3851c27e8.png)

**Comment by YanVugenfirer** [comment]
@sb-ntnx Hi Sergey

Based on the previous comment, do you think that the issue is specific to AHV?

Thanks,
Yan.

**Comment by sb-ntnx** [comment]
Hi @YanVugenfirer, 

Unfortunately, I've not had time to look into the issue yet. I hope to have a further look later this week. 

BR,
Sergey

**Comment by jborean93** [comment]
Just thought I would share that I am encountering this issue with using the `virtio` network adapter in VirtualBox. Interestingly enough QEMU isn't affected. Using the workaround mentioned in https://github.com/virtio-win/kvm-guest-drivers-windows/issues/583#issuecomment-845242351 by settings `Offload.Tx.Checksum` to `Disabled` does mitigate the issue in VirtualBox for me. A way to set this through PowerShell is:

```powershell
Get-NetAdapter |
    Where-Object DriverFileName -eq 'netkvm.sys' |
    Set-NetAdapterAdvancedProperty -RegistryKeyword 'Offload.TxChecksum' -RegistryValue 0
```

**Comment by sb-ntnx** [comment]
Sorry for the delay. It has been fairly busy recently, so I've not had time yet to throughly investigate the issue. However, I tested with qemu-6.0.0 (with AHV) and got the same behavior. 

BR,
Sergey

**Comment by YanVugenfirer** [comment]
@sb-ntnx What's the host networking configuration? 
Also, are you using vhost?

According to @leidwang on QEMU\KVM with vhost and Linux bridge, there is no problem.

**Comment by sb-ntnx** [comment]
The qemu command line is: 
`-netdev tap,fd=29,id=hostnet0,vhost=on,vhostfd=31
-device virtio-net-pci,rx_queue_size=256,netdev=hostnet0,id=net0,mac=50:6b:8d:c0:ca:18,bus=pci.0,addr=0x3,bootindex=3`

By default, tap interfaces are added to ovs bond, but I tried with a plain Linux switch between two VMs running on the same host:
`[root@svc1-3 ~]# brctl addbr HCK
[root@svc1-3 ~]# ip link set HCK up
[root@svc1-3 ~]# virsh domiflist 017d4bf9-5a49-4bdf-ad39-7d108a0cffc9
Interface  Type       Source     Model       MAC
-------------------------------------------------------
tap1       ethernet   -          virtio      50:6b:8d:9b:59:e6

[root@svc1-3 ~]# virsh domiflist 57b53584-6eb7-46c3-bc0c-a235184c88e7
Interface  Type       Source     Model       MAC
-------------------------------------------------------
tap0       ethernet   -          virtio      50:6b:8d:c0:ca:18
tap2       ethernet   -          e1000       50:6b:8d:e3:5f:f3

[root@svc1-3 ~]# ovs-vsctl del-port tap0
[root@svc1-3 ~]# ovs-vsctl del-port tap1
[root@svc1-3 ~]# brctl addif HCK tap0
[root@svc1-3 ~]# brctl addif HCK tap1
`

However, what I noticed is that DNS resolution may work and you actually need to examine the packets on the receiving server if the checksum is correct: 
Screenshot 1: DNS request coming when running `ipconfig /flushdns && ping -n 1 t.co`
![image](https://user-images.githubusercontent.com/56913349/122565666-06c8a400-d047-11eb-8578-6618add05c4c.png)

Screenshot 2: DNS request, when trying to resolve t.co using nslookup.exe: 
![image](https://user-images.githubusercontent.com/56913349/122565828-3677ac00-d047-11eb-9d53-8afef2367477.png)

In the lab test with Linux bridge, I noticed that for DNS lookups initiated by `ping.exe -n 1 t.co`, if we compare what leaves the client VM and what arrives at the DHCP server, we see that the packet is bigger and I don't know how to explain it: 
![image](https://user-images.githubusercontent.com/56913349/122566816-43e16600-d048-11eb-8b84-d961fd914eb4.png)

If the lookup is done via nslookup.exe, then we see the same packet size on the server and client: 
![image](https://user-images.githubusercontent.com/56913349/122567186-b2bebf00-d048-11eb-9f64-c5e4428886e2.png)

**Comment by YanVugenfirer** [comment]
@sb-ntnx Can you please specify what was the Linux kernel version?

**Comment by YanVugenfirer** [comment]
@jborean93 What was the host OS when you ran VirtualBox?

**Comment by jborean93** [comment]
> @jborean93 What was the host OS when you ran VirtualBox?

It was Fedora 34, the current kernel is `5.12.12-300.fc34.x86_64` but that may be slightly different due to a dnf update in the recent past.

**Comment by sb-ntnx** [comment]
@YanVugenfirer: The latest kernel of AHV that I tested was based off 5.4.109

**Comment by leidwang** [comment]
Tested this issue on Win2022,still not hit the problem.
Host:
kernel-4.18.0-321.el8.x86_64
qemu-kvm-6.0.0-22.module+el8.5.0+11695+95588379.x86_64
![latest_kernel_583](https://user-images.githubusercontent.com/63003119/124697664-c6f42e80-df19-11eb-9fa6-ee363e968be4.png)

**Comment by leidwang** [comment]
Still not hit this issue with kernel-5.13.
Env:
kernel-5.13.0-0.rc7.51.el9.x86_64
qemu-kvm-6.0.0-7.el9.x86_64
Guest(Win2022):
Windows_InsiderPreview_Server_vNext_en-us_20344.iso
virtio-win-prewhql-0.1-202.iso
![583_kernel5 13](https://user-images.githubusercontent.com/63003119/124939022-60ffc800-e03b-11eb-9412-13ca971116fb.png)

**Comment by sb-ntnx** [comment]
@leidwang, thank you for the tests.
@YanVugenfirer: could you point me to the piece of code where csums are calculated?

**Comment by YanVugenfirer** [comment]
@sb-ntnx 
Some explanation regarding checksum offload. In general we will offload checksum calculation to the host. But in the case of TCP fragmentation enabled and IP checksum turned on, we cannot offload both (virtio-net-hdr has only place for one checksum metadata).
So you can follow the "trail" from ```bool CNB::BindToDescriptor(CTXDescriptor &Descriptor)``` in ParaNdis_TX.cpp -> ```void CNB::PrepareOffloads(virtio_net_hdr *VirtioHeader, PVOID IpHeader, ULONG EthPayloadLength, ULONG L4HeaderOffset) const``` -> SetupLSO or DoIPHdrCSO and then code in sw_offload.cpp (through ParaNdis_CheckSumVerifyFlat) ```tTcpIpPacketParsingResult ParaNdis_CheckSumVerify(
                                                tCompletePhysicalAddress *pDataPages,
                                                ULONG ulDataLength,
                                                ULONG ulStartOffset,
                                                ULONG flags,
                                                BOOLEAN verifyLength,
                                                LPCSTR caller)```

**Comment by sb-ntnx** [comment]
@YanVugenfirer: I guess we can close this bug out as the issue is understood and the PR is merged into the upstream? 

@olljanat: We will make this change as part of the VirtIO 1.1.7 suite. Cannot commit to the release date, but we will try to expedite it.

**Comment by YanVugenfirer** [comment]
> @YanVugenfirer: I guess we can close this bug out as the issue is understood and the PR is merged into the upstream?
> 
Yes, I will close it. Thanks a lot for PR.

**Comment by nchevsky** [comment]
FYI, @sb-ntnx's [fix](https://github.com/virtio-win/kvm-guest-drivers-windows/pull/607/commits/4ecff566fced55d970b0d573c2c0183a103ab4c9) is included in Fedora's [0.1.208-1 release](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/archive-virtio/virtio-win-0.1.208-1/) of the `virtio-win` drivers. 🎉

**Comment by YanVugenfirer** [comment]
@zx2c4 Thanks.

**Comment by wioxjk** [comment]
Can confirm that this is an issue in Nutanix AHV with Windows Server 2022.
I have opened up a ticket with them regarding this.

**Comment by sb-ntnx** [comment]
@wioxjk as it's been mentioned in https://github.com/virtio-win/kvm-guest-drivers-windows/issues/583#issuecomment-887554827, the fix was delivered in Nutanix VirtIO 1.1.7.
